### DISCUSSION TITLE
OPTIONS method on a listview calls `get_object()` which fails an assertion

### DISCUSSION THREAD

**Discussion by stiffupperlip** [encode/django-rest-framework#7757] [Potential Issue]
When I send OPTIONS request to this view, it tries to call `generics.GenericAPIView.get_object()` method which raises assertion error, because there is no `pk` argument in url pattern, which is not needed because this is a list view.

URL pattern:

```
url(r'api/services/?$', views.ServiceView.as_view()),
url(r'api/services/(?P<pk>\d+)/?$', views.ServiceDetailView.as_view()), # this is another view that `OPTIONS` work fine for.
```

View class:

```
class ServiceView(generics.GenericAPIView):
    queryset = models.Service.objects.all()
    serializer_class = serializers.ServiceSerializer
    search_fields = ('name',)
    filter_fields = ('id', 'name')
    ordering_fields = '__all__'
    ordering = ('id',)
```

When I send this request:
`http --auth username:password options localhost:8000/api/services`

Here's the traceback:

```
api.views ERROR Exception "AssertionError": "Expected view ServiceView to be called with a URL keyword argument named "pk". Fix your URL conf, or set the `.lookup_field` attribute on the view correctly."
Traceback (most recent call last):
  File "/home/a/.virtualenvs/ch/local/lib/python2.7/site-packages/rest_framework/views.py", line 463, in dispatch
response = handler(request, *args, **kwargs)
  File "/home/a/.virtualenvs/ch/local/lib/python2.7/site-packages/rest_framework/views.py", line 477, in options
data = self.metadata_class().determine_metadata(request, self)
  File "/home/a/.virtualenvs/ch/local/lib/python2.7/site-packages/rest_framework/metadata.py", line 69, in determine_metadata
actions = self.determine_actions(request, view)
  File "/home/a/.virtualenvs/ch/local/lib/python2.7/site-packages/rest_framework/metadata.py", line 88, in determine_actions
view.get_object()
  File "/home/a/.virtualenvs/ch/local/lib/python2.7/site-packages/rest_framework/generics.py", line 93, in get_object
(self.__class__.__name__, lookup_url_kwarg)
AssertionError: Expected view ServiceView to be called with a URL keyword argument named "pk". Fix your URL conf, or set the `.lookup_field` attribute on the view correctly.
[14/Mar/2016 17:30:21] "OPTIONS /api/services HTTP/1.1" 500 33
```

`OPTIONS` method was used to work when I subclassed `APIView` for views, but now that I'm subclassing `generics.GenericAPIView`, It fails.

Django REST framework 3.3.3, Django 1.8.11, Python 2.7.6, Ubuntu 14.04

**Comment by xordoquy** [comment]
Which is strange because `get_object` will only be called if it exists _and_ if the method is `PUT`.
I think the next step here would be an isolated test case that highlights the issue.

**Comment by narrowfail** [comment]
Hello, same error over here! When I enable OPTIONS and PUT, I get an AssertionError. Example over here:

```
class EmployeeModelSerializer(ModelSerializer):
    responsible = SameCompanyEmployeeField(allow_null=True, allow_empty=True)

    def validate_responsible(self, data):
        if data and (data.pk == self.instance.pk):
            raise ValidationError('Self reference not allowed!')
        return data

    class Meta:
        model = EmployeeModel
        exclude = ('organization',)


class EmployeeModelViewSet(ModelViewSet):
    serializer_class = EmployeeModelSerializer
    queryset = EmployeeModel.objects.all()
    allowed_methods = ('GET', 'POST', 'PUT', 'HEAD', 'DELETE', 'OPTIONS')
    lookup_field = 'id'
    filter_fields = ('first_name', 'last_name', 'role', 'title', 'abilities',
                     'responsibilities', 'training_needs',)

```

Exception Type: AssertionError at /api/v1/employee/
Exception Value: Expected view EmployeeModelViewSet to be called with a URL keyword argument named "id". Fix your URL conf, or set the `.lookup_field` attribute on the view correctly.`

I get the same traceback. I think you guys missed adding an AssertionError in the except in /rest_framework/metadata.py line 88 (but I don't know the framework well).

Thanks,
Guillermo

**Comment by narrowfail** [comment]
@xordoquy did you get this info?

**Comment by beltiras** [comment]
I have this same problem. OPTIONS on some object works fine. OPTIONS on the list fails in the same way. My viewset is very comparable to @narrowfail. If I remove the lookup_field argument everything works as intended.

**Comment by narrowfail** [comment]
@xordoquy can you check this Issue?
